docs/tools/build_rules.md

```diff
@@ -0,0 +1,118 @@
+## Mbed OS Build Rules
+
+The Mbed OS Build Tools scan your project for source files every time you compile. This document describes the rules that the build tools use to decide which source files to include in each build. The Mbed OS Build Tools use the [target configuration](mbed_targets.md) found in `targets.json`, `mbed_app.json` and `mbed_lib.json` as input to these rules.
+
+The Build Tools include every source file found in the project unless it is in a label directory, it is in a test directory or it matches a pattern in an `.mbedignore file`.
```

Comment:
Suggest identifying what constitutes a source file... especially concerned about *.asm/*.s for assembly

Reply:
That particular thing might change soon. I can document what we currently consider a source file.

Reply:
Added by f77a96f